controller/tenant.go

```diff
@@ -198,13 +198,16 @@ func (c *TenantController) loadUserTenantConfiguration(ctx context.Context, conf
 			return config, err
 		}
 		resp, err := authClient.ShowUser(ctx, auth.ShowUserPath(), nil, nil)
+		defer resp.Body.Close()
+
 		if err != nil {
 			log.Error(ctx, map[string]interface{}{"auth_url": auth.ShowUserPath()}, "unable to get user info")
 			return config, errs.Wrapf(err, "failed to GET url %s due to error", auth.ShowUserPath())
 		}
 		if resp.StatusCode < 200 || resp.StatusCode > 300 {
 			return config, fmt.Errorf("failed to GET url %s due to status code %d", resp.Request.URL, resp.StatusCode)
```

Comment:
You need to read the body here too because otherwise it may leak FD even if you close the body :)
General rule for making sure we don't have a FD leak - If err!=nil then read the body (once) and close it.

Reply:
ok, let me move this line just after `js, err := simplejson.NewFromReader(resp.Body)`, then. I assume this will respond to the golden rule you explained, right ?

Reply:
If response is not a JSON (if it's not 200) then simplejson.NewFromReader(resp.Body) will fail I guess. So, instead do the following:
```
defer resp.Body.Close()
if err != nil {
   // return error
}
body := ioutil.ReadAll(resp.Body)
if resp.StatusCode < 200 || resp.StatusCode > 300 {
   // return error
}

// Unmarshal JSON from body.
```

Or

```
defer resp.Body.Close()
if err != nil {
   // return error
}
if resp.StatusCode < 200 || resp.StatusCode > 300 {
   ioutil.ReadAll(resp.Body)
   // return error
}

js, err := simplejson.NewFromReader(resp.Body)
```

Or simply use `defer rest.CloseResponse(resp.Body)` from https://github.com/fabric8-services/fabric8-auth/blob/master/rest/url.go#L46 ;)

Reply:
ok, thanks for the explanations, @alexeykazakov. Am I wrong or does 
````
func CloseResponse(response *http.Response) {
	ioutil.ReadAll(response.Body)
	response.Body.Close()
}
````
reads the body but does not return it ? I thought we should read it once only ? Or at least once ? (In Java you cannot read the streams multiple times)

Reply:
fixed in https://github.com/fabric8-services/fabric8-tenant/pull/495/commits/06ade20f0fd51f6c2508612e6cff20f915b9eaa5

Reply:
hopefully ;)

Reply:
That's right. It reads the stream to make sure it's empty and then closes the body. If stream is already empty (body has been read) that's OK. It doesn't do any harm.